File: 6_linked_list_and_arrays/3_3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = set()
        nums.sort()
        n = len(nums)
        for i in range(n):
            seen = set()
            for j in range(i + 1, n):
                required_num = 0 - nums[i] - nums[j]
                if required_num in seen:
                    res.add((nums[i], nums[j], required_num))
                seen.add(nums[j])
        return res


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = set()
        n = len(nums)
        for i in range(n):
            left, right = i + 1, n - 1
            while left < right:
                curr_sum = nums[i] + nums[left] + nums[right]
                if not curr_sum:
                    res.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1
                elif curr_sum > 0:
                    right -= 1
                else:
                    left += 1
        return res

File: 07_linked_list_and_arrays/3_3sum.rs
use std::collections::HashSet;

impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums = nums;
        let mut res: HashSet<(i32, i32, i32)> = HashSet::new();
        nums.sort();
        let n = nums.len();

        for i in 0..n {
            let (mut l, mut r) = (i + 1, n - 1);
            while l < r {
                let curr = nums[i] + nums[l] + nums[r];
                if curr == 0 {
                    res.insert((nums[i], nums[l], nums[r]));
                    l += 1;
                    r -= 1;
                } else if curr > 0 {
                    r -= 1;
                } else {
                    l += 1;
                }
            }
        }

        let mut return_value: Vec<Vec<i32>> = vec![];
        for item in res {
            return_value.push(vec![item.0, item.1, item.2]);
        }
        return_value
    }
}
